chirp/drivers/baofeng_digital.py

```diff
@@ -0,0 +1,492 @@
+            return self._pack_ctcss(ctcss_freq)
+        elif mode == 'DTCS':
+            return self._pack_dtcs(dtcs_code, dtcs_polarity)
+        else:
+            return NO_TONE
+
+    def set_memory(self, mem):
+        raw = self._memobj.memory[mem.number - 1]
+        if mem.empty:
+            raw.set_raw(b'\xFF' * 16)
+            return
+
+        raw.set_raw(b'\x00' * 16)
+
+        raw.rx_freq = mem.freq // 10
+        if mem.duplex == '':
+            raw.tx_freq = mem.freq // 10
+        elif mem.duplex == 'split':
+            raw.tx_freq = mem.offset // 10
+        elif mem.duplex == '-':
+            raw.tx_freq = (mem.freq - mem.offset) // 10
+        elif mem.duplex == '+':
+            raw.tx_freq = (mem.freq + mem.offset) // 10
+        else:
+            raise errors.RadioError('Unsupported duplex mode %r' % mem.duplex)
+
+        raw.narrow = mem.mode in ('NFM', 'DN')
+        raw.digital = mem.mode in ('DIG', 'DN')
+        try:
+            raw.high_power = self.POWER_LEVELS.index(mem.power)
+        except ValueError:
+            raw.high_power = 1
+        raw.scan = mem.skip != 'S'
+
+        if raw.digital:
+            raw.tx_tone = NO_TONE
+            raw.rx_tone = NO_TONE
+        elif mem.tmode == 'Tone':
+            raw.tx_tone = self._pack_ctcss(mem.rtone)
+            raw.rx_tone = NO_TONE
+        elif mem.tmode == 'TSQL':
+            raw.tx_tone = raw.rx_tone = self._pack_ctcss(mem.ctone)
+        elif mem.tmode == 'DTCS':
+            # According to the documentation we should use ".rx_dtcs" because
+            # "has_rx_dtcs" is set. However, the code does not match the docs,
+            # and we have to use ".dtcs".
+            raw.tx_tone = self._pack_dtcs(mem.dtcs, mem.dtcs_polarity[0])
+            raw.rx_tone = self._pack_dtcs(mem.dtcs, mem.dtcs_polarity[1])
+        elif mem.tmode == 'Cross':
+            tx_tmode, rx_tmode = mem.cross_mode.split('->', 2)
+            raw.tx_tone = self._pack_squelch(
+                tx_tmode, mem.rtone, mem.dtcs, mem.dtcs_polarity[0])
+            raw.rx_tone = self._pack_squelch(
+                rx_tmode, mem.ctone, mem.rx_dtcs, mem.dtcs_polarity[1])
+
+        raw.bcl_disabled = 1
+        for item in mem.extra:
+            if item.get_name() == 'bcl':
+                raw.bcl_disabled = not item.value
+            elif item.get_name() == 'hopping' and not raw.digital:
+                raw.hopping = item.value
+            elif item.get_name() == 'encryption_key' and raw.digital:
+                raw.encryption_key = item.value
+
+
+@directory.register
+class BaofengW31D(BaofengDigital):
+    """Baofeng W31D"""
+    MODEL = 'W31D'
+    VALID_BANDS = [(400000000, 480000000)]
+    POWER_LEVELS = [
+        chirp_common.PowerLevel("Low", watts=0.5),
+        chirp_common.PowerLevel("High", watts=2.00),
+    ]
+    _serial_id = 0x04
+    _proto = PROTO_D
+
+
+# Disabled because UNTESTED, though it should work
+# @directory.register
```

Comment:
Let's not put this into the tree like this and instead attach a .py file to the (or a new) tracker issue  and let people test it as a module.